### ISSUE TITLE
Enhancement: Less obfuscated versions of some winners

### ISSUE THREAD

**Issue by SirWumpus** [ioccc-src/temp-test-ioccc#2513]
I was reviewing how my past entries currently appear in the updated site.

All my winning entries are on GitHub, have spoilers, some have been cleaned up to C89 versions (ae, ag).  All of them should work.  Much of the documentation has been converted to Markdown (not the .hint files).   Not sure if you want to reference these repositories, copy the formatted documentation, etc.

Anyway I provide these links to the material:

* 1991 https://github.com/SirWumpus/ioccc-ae/91 (top level provides history and updated version)
* 1992 https://github.com/SirWumpus/ioccc-am
* 1993 https://github.com/SirWumpus/ioccc-ag
* 2004 https://github.com/SirWumpus/ioccc-hibachi
* 2005 https://github.com/SirWumpus/ioccc-mynx (original http: version)
* 2015 https://github.com/SirWumpus/ioccc-npdif

**Comment by xexyl** [comment]
Thanks! Landon told me that there shouldn't be any hidden spoilers nowadays .. I can help out here.

**Comment by xexyl** [comment]
Actually .. I was going to do this now but I wonder something. Landon, do you think the spoiler files should be included in the website or just have them linked to from the index.html file ? I can do either one but including the files obviously means a manifest update. I'll hold off until you let me know what you prefer, @lcn2, and go from there.

**Comment by SirWumpus** [comment]
In my repos, I have the original entry at the root with a `spoiler/` directory.

Would you link to the repo root or to the spoiler within the repo?  I recommend the repo root, because the owner may restructure it in the future breaking links.

**Comment by xexyl** [comment]
> I originally proposed the idea of including spoilers a few years ago (didn't want to get hit by a ebike without having shared the originals), which were available in some places on `ioccc.org`.
> 

Well I'm glad you didn't get hit by an e-bike .. or any other kind!

> I think if the author is maintaining the code on GitHub, GitLab, ... then a suitable link in the repository is probably sufficient (fewer changes), but of the author has abandoned the code and left spoilers, then a `spoiler/` directory seems apropos and I think at this stage you would have accounted for those cases in a manifest.
> 

That's probably good enough but not sure what Landon thinks. I'd agree it's enough though. And if the files are not added then the manifest does not need to be updated.

I'd say that maybe the only reason to include files is in case the repo is deleted. That seems most unlikely but the problem with including the files too is if they change (as you say in your next comment). 
> I did see place last night `source/` directories? Where these spoilers or part of the original submission; if spoiler then I think an explicit name like `spoiler/` is clearer as to purpose.

Well as for that one entry had a spoiler/ subdirectory but it was moved to the directory itself. This was in 2020: it had been a zip file which was password protected but the password was no longer known (wiki article that was changed). I had it from the winner archive though so I just added the files.

But I'm not sure if a subdirectory is necessary or not. If it is then `spoiler/` sounds like a good name to me too.

**Comment by xexyl** [comment]
> In my repos, I have the original entry at the root with a `spoiler/` directory.
> 

I noticed that. The first link you had I think was broken but I corrected it in the browser and got to it.

> Would you link to the repo root or to the spoiler within the repo? I recommend the repo root, because the owner may restructure it in the future breaking links.

Good point indeed.

**Comment by SirWumpus** [comment]
> The first link you had I think was broken but I corrected it in the browser and got to it.

You referring to the [https://github.com/SirWumpus/ioccc-ae](https://github.com/SirWumpus/ioccc-ae/91)?  That one is rather specially organised, because there was a some derivative work that followed after 1991.

**Comment by xexyl** [comment]
> > The first link you had I think was broken but I corrected it in the browser and got to it.
> 
> You referring to the [https://github.com/SirWumpus/ioccc-ae](https://github.com/SirWumpus/ioccc-ae/91)? That one is rather specially organised, because there was a some derivative work that followed after 1991.

Yes that one .. but the link is 404. https://github.com/SirWumpus/ioccc-ae/tree/master seems good.

**Comment by SirWumpus** [comment]
That doesn't make sense.  It works for me and DuckDuckGo search lists the repo URL in the results ( https://duckduckgo.com/?t=ffab&q=howe+ioccc+ae ).  I would avoid reference `.../tree/master` or `.../tree/main` or whatever GitHub will use to point into the source tree.

**Comment by xexyl** [comment]
> That doesn't make sense.  It works for me and DuckDuckGo search lists the repo URL in the results ( https://duckduckgo.com/?t=ffab&q=howe+ioccc+ae ).  I would avoid reference `.../tree/master` or `.../tree/main` or whatever GitHub will use to point into the source tree.

I thought it was odd too but clicking on it got 404. I would also avoid the last part but I just copy pasted it from Safari without worrying about changing it.

**Comment by SirWumpus** [comment]
OK. Tested in Firefox and Vivaldi.  The top level URL https://github.com/SirWumpus/ioccc-ae works as expected, while subdirectory https://github.com/SirWumpus/ioccc-ae/91 does not (404).  From the top level README, GitHub converts the relative link to something like https://github.com/SirWumpus/ioccc-ae/blob/master/91 so using internal links into a repo can vary.

**Comment by lcn2** [comment]
The IOCCC web site has shifted to a more educational format: preferring to explain  entries rather than hiding things under the guise of such information being a "spoiler". 

As a result, content that was "hidden" is now being made available.  For example, stuff that was rot13 are now plaintext. 

Instead of a "can you guess", the IOCCC favors people being able to learn about an IOCCC winning entry. 

If there is a de-obfuscated version of an IOCCC winning entry, adding such code would be welcome.  How to add a de-obfuscated version is up for discussion: one way is the provide an _alt_ code version and reference it in the `README.md` file. Another would be to add it as some sort of `prog.deobfuscated.c` file and again reference it in the `README.md` file. There may be other ways as well. 

We would have the _alt_ method (as `prog.alt.c` or as `prog.alt2.c` in case `prog.alt.c` already exists) and then referencing the de-obfuscated code in the `README.md` as something the reader might wish to explore.

Anyone who would care to provide de-obfuscated code would be thanked and credited in the `thanks-for-the-help.md` file with our thanks. 🙏 

So, @SirWumpus, if you have de-obfuscated code versions, we would be happy to add them into the collection with our thanks. 🙏

**Comment by SirWumpus** [comment]
I've always tried to educate with each of my entries.

I would reserve `*.alt.c` or alt anything for literal alternative versions, typically compiler bug fixes or minor change to the original entry.

I still think the original developer's code (and any support tools they provide) go in a `spoiler/` or if you prefer `directors_cut/` or similar directory to denote that it was not part of the original submission, but provides more details, like a commented version, full source, extra docs and reference material, etc.

**Comment by xexyl** [comment]
> The IOCCC web site has shifted to a more educational format: preferring to explain entries rather than hiding things under the guise of such information being a "spoiler".
> 
> As a result, content that was "hidden" is now being made available. For example, stuff that was rot13 are now plaintext.

Yes and also `uuencode`d things .. I undid both rot13 things (except in one case where it wasn't rot13 but a different rotation - if you recall) and `uuencode`. I did not undo the 2020/ferguson2 one of course: that is a way to experiment with the entry and it also has some good fun in it too! And that's another reason we don't always do it - if it's part of the entry it should not be done (though I guess if by [some chance](https://rationalwiki.org/wiki/Zero#Dividing_by_zero) [0] one won it would not be that way).

> 
> Instead of a "can you guess", the IOCCC favors people being able to learn about an IOCCC winning entry.
> 

This might be something that can't say as it pertains to judging but I guess that it is not actually going to decrease the chance of winning if this is not done? (Whether it would increase if it is done I know you would not say and I would not want you to say it either).

> If there is a de-obfuscated version of an IOCCC winning entry, adding such code would be welcome. How to add a de-obfuscated version is up for discussion: one way is the provide an _alt_ code version and reference it in the `README.md` file. Another would be to add it as some sort of `prog.deobfuscated.c` file and again reference it in the `README.md` file. There may be other ways as well.
> 

I refer to that 'way to go about it' below. We often do `alt` but in at least one or two cases there's the name of the file given by the author say on their website (like with one entry of Bellard).

> We would have the _alt_ method (as `prog.alt.c` or as `prog.alt2.c` in case `prog.alt.c` already exists) and then referencing the de-obfuscated code in the `README.md` as something the reader might wish to explore.
> 

That's one way of doing it. I think I did that in a few cases but I'm not sure. In some cases it was already done - code submitted with the entry itself was turned into an alt version.

> Anyone who would care to provide de-obfuscated code would be thanked and credited in the `thanks-for-the-help.md` file with our thanks. 🙏
> 

Ha. I did do that for a few .. maybe. I'm not sure. But I found it on the respective author's repo (or in one case website). 

> So, @SirWumpus, if you have de-obfuscated code versions, we would be happy to add them into the IOCCC collection with our thanks. 🙏

I can add those .. unless you ish to, @SirWumpus ? Let me know and I'd be happy to do it. On second thought .. I guess you don't have macOS so you can't update the manifest. Okay I'll do that in the next few days I hope. 

A question for you Landon is: where should they be? A subdirectory called `spoiler/` or in the entry's directory? If in the entry's directory what should they be called? Perhaps the name of the files in the particular directory? I guess we should do the alt method but the question is how many alt versions should there be. This is hypothetical but if there were 42 revisions we would not want all the way up to .alt41 (or whatever we have used .. can't remember right now - too tired and can't be bothered checking). In at least one case I called the file the name the author gave (those I found - didn't touch the ones that the author submitted originally).

Of course if @SirWumpus prefers to do it one of us can always fix the manifest. But might just be easier if I do it so the questions above still apply I think?


[0] Siri really has told that joke. I have a screenshot of it .. alas I tried it again some years ago and no luck :(

**Comment by lcn2** [comment]
> > We would have the _alt_ method (as `prog.alt.c` or as `prog.alt2.c` in case `prog.alt.c` already exists) and then referencing the de-obfuscated code in the `README.md` as something the reader might wish to explore.
> 
> That's one way of doing it. I think I did that in a few cases but I'm not sure. In some cases it was already done - code submitted with the entry itself was turned into an alt version.

The focus from those initial readings the website is on the entry `index.html` file (as generated from the `README.md`) file.

If the _alt_ section, if there is a de-obfuscated version of `prog.c`, then the entry `index.html` file (as generated from the `README.md`) file should mention that file as being a de-obfuscated version (formerly "_sppiler_").

Those wishing to just "download, compiler and dive-into the code" can just do a `make` and starting reading `prog.c` as the `prog.c` code will **never** be a de-obfuscated version. 

There are very special cases where `make` builds an "_alt_" version of the code, but that case the "_alt_" code is **NOT** a de-obfuscated version  but rather version that is able to compile or otherwise fixes something "fatal" or "nearly fatal" with the original code.

Moreover `prog.orig.c` remains the original winning source (or as close we have to the original), and again **never** be a de-obfuscated version.

> > Anyone who would care to provide de-obfuscated code would be thanked and credited in the `thanks-for-the-help.md` file with our thanks. 🙏
> 
> Ha. I did do that for a few .. maybe. I'm not sure. But I found it on the respective author's repo (or in one case website).

👍

> > So, @SirWumpus, if you have de-obfuscated code versions, we would be happy to add them into the IOCCC collection with our thanks. 🙏
> 
> I can add those .. unless you wish to, @SirWumpus ? Let me know and I'd be happy to do it. On second thought .. I guess you don't have macOS so you can't update the manifest. Okay I'll do that in the next few days I hope.

Well @SirWumpus, I would recommend working with @xexyl to contribute your de-obfuscated code.  If you provide @xexyl with the URLs of the code, and any additional commentary that might prove educational, @xexyl can go through the steps of adding your contributions with our **THANKS** and appreciation!

> A question for you Landon is: where should they be? A subdirectory called `spoiler/` or in the entry's directory? If in the entry's directory what should they be called? Perhaps the name of the files in the particular directory? I guess we should do the alt method but the question is how many alt versions should there be. This is hypothetical but if there were 42 revisions we would not want all the way up to .alt41 (or whatever we have used .. can't remember right now - too tired and can't be bothered checking). In at least one case I called the file the name the author gave (those I found - didn't touch the ones that the author submitted originally).
> 
> Of course if @SirWumpus prefers to do it one of us can always fix the manifest. But might just be easier if I do it so the questions above still apply I think?
> 
> [0] Siri really has told that joke. I have a screenshot of it .. alas I tried it again some years ago and no luck :(

Adding a spoiler sub-directory with code to compile and clean and clobber adds some more complicatation the build system, so we would prefer to not use such a subdirectory mechanism.

**Comment by SirWumpus** [comment]
All the spoiler code and any extra stuff are in a `spoiler/` for each repo [from above](https://github.com/ioccc-src/temp-test-ioccc/issues/2513#issue-2382180811).

> Adding a spoiler sub-directory with code to compile and clean and clobber adds some more complicatation the build system, so we would prefer to not use such a subdirectory mechanism.

I'd suggest NOT building (or cleaning) `spoiler/` subdirectories (ohh, just remembered `contrib/` is a fav of many, might be better than `spoiler/`) as part of the build system.  These are supplemental to the winner and provided for learning.  Leave them to the viewers to play with IMO.  I wouldn't fuss with the extra stuff- just provide as-is.

**Comment by xexyl** [comment]
Excellent comment Landon and great idea about the FAQ. I think that would be a good idea. The question is do you want to do that or should I?

I will reply to your comment in more detail tomorrow when at the laptop. I will be going to bed shortly.

**Comment by lcn2** [comment]
> Excellent comment Landon and great idea about the FAQ. I think that would be a good idea. The question is do you want to do that or should I?
> 
> I will reply to your comment in more detail tomorrow when at the laptop. I will be going to bed shortly.

We can draft an initial FAQ for you to modify as needed ... doing that now ..

**Comment by xexyl** [comment]
> > Excellent comment Landon and great idea about the FAQ. I think that would be a good idea. The question is do you want to do that or should I?
> > I will reply to your comment in more detail tomorrow when at the laptop. I will be going to bed shortly.
> 
> We will draft an initial FAQ for you to modify as needed ... doing that now ..
> 
> ## UPDATE 0
> With commit [4d37308](https://github.com/ioccc-src/temp-test-ioccc/commit/4d37308d120cca180a7c0dee3e128c145dd7c907) added a draft of [FAQ 5.8](https://ioccc-src.github.io/temp-test-ioccc/faq.html#deobfuscated).
> 
> ### QUESTION
> Should we use the term "_deobfuscated_" or "_de-obfuscated_"?
> 
> We used "_deobfuscated_" below.
> 
> ## UPDATE 1d
> With commit [4d6009b](https://github.com/ioccc-src/temp-test-ioccc/commit/4d6009b58fa56e966818fc4b42d8741ec4f64e73), and commit [3aec80f](https://github.com/ioccc-src/temp-test-ioccc/commit/3aec80f057cf4863a41fba40e2039293694aead1), and commit [85c9916](https://github.com/ioccc-src/temp-test-ioccc/commit/85c991620f605d3905f386d52fe3ed3a2a32fa85), and commit [5dd3311](https://github.com/ioccc-src/temp-test-ioccc/commit/5dd3311e9ef36b4570cd972b6d1ec327ae7a4c78): we shifted away from "_spoiler_" language and filenames and to "_deobfuscation_" information and filenames: as per [FAQ 5.8](https://ioccc-src.github.io/temp-test-ioccc/faq.html#deobfuscated).
> 
> So all existing "_spoiler_" usage have been converted into "_deobfuscated_", although one may wish to check out and correct these edits.

Yes I think some might be wrong indeed. One is in 2020/ferguson1 in fact. It's not really about de-obfuscation (well some of it is but it's not strictly it). It certainly is not an unobuscated version! It's just notes on how it's obfuscated so it might be better to call it (in this case) `obfuscation.{md,html}` ? I like that. And I'm not sure about some other files either ... The ones with JavaScript that type out the code: I'm not sure. Those by I think Don Yang. For instance what do you think the name of 2019/yang/deobfuscation.html should be? In this case it might work as it shows how it was developed.

On the other hand the file: 2018/algmyr/deobfuscation.png doesn't seem like the right name, somehow.

That being said I do think 'deobfuscated' is better than 'de-obfuscated'. But then again unobfuscated might also be valid depending on the context ... I don't know. What I do know though is that in some cases the way you put it is definitely not right.

The question is what else has to be done with the changing of the file names. That I'm not sure other than updating the manifest. In the case of 2020/ferguson1 it's referenced in at least some files so I can just use `sgit(1)` on those files .. but not so simple on all of them, maybe.

In any case I will change 2020/ferguson1 now and then we can discuss the others later. I hope to look at the guidelines in a bit too ...

**Comment by xexyl** [comment]
.. actually I'm not sure that's the right name for 2020/ferguson1 either. The name spoilers works well but obfuscation does not work as well as it's not just about obfuscation. It has as a table of contents:

```-   [How does it work?](#works)
-   [How does the snake movement work?](#movement)
-   [Drawing &#x28;manual&#x29; and computer playing &#x28;automatic&#x29; modes](#manual-automatic)
-   [Collision detection](#collision)
    *   [Cannibal collision detection](#cannibalcollision)
    *   [Collision detection when resizing the window of the game](#resizing)
-   [Obfuscation](#obfuscation)
-   [Skinning the snake &#x28;i.e., decreasing the iocccsize&#x29;](#skinning)
-   [A few more interesting size optimisations](#size)
```

so `deobfuscation` and variants are definitely wrong but I think obfuscation an variants are also wrong. The problem is that since it does talk about obfuscation (rather than deobfuscation) AND the other things it can't simply be named 'obfuscation'. Something like `details` (for implementation details) doesn't work that well either because it has the obfuscation part and more importantly other files also talk about these things ...

So I don't know.

**Comment by xexyl** [comment]
> .. actually I'm not sure that's the right name for 2020/ferguson1 either. The name spoilers works well but obfuscation does not work as well as it's not just about obfuscation. It has as a table of contents:
> 
> ```
> -   [How does the snake movement work?](#movement)
> -   [Drawing &#x28;manual&#x29; and computer playing &#x28;automatic&#x29; modes](#manual-automatic)
> -   [Collision detection](#collision)
>     *   [Cannibal collision detection](#cannibalcollision)
>     *   [Collision detection when resizing the window of the game](#resizing)
> -   [Obfuscation](#obfuscation)
> -   [Skinning the snake &#x28;i.e., decreasing the iocccsize&#x29;](#skinning)
> -   [A few more interesting size optimisations](#size)
> ```
> 
> so `deobfuscation` and variants are definitely wrong but I think obfuscation an variants are also wrong. The problem is that since it does talk about obfuscation (rather than deobfuscation) AND the other things it can't simply be named 'obfuscation'. Something like `details` (for implementation details) doesn't work that well either because it has the obfuscation part and more importantly other files also talk about these things ...
> 
> So I don't know.
> 
> .. and I see some grammar issues in that table of contents: odd. I don't recall having added a comma after `i.e.` : I wouldn't. But maybe I did by accident (ah: just in the table of contents, not in the title itself .. which is expected).

Now as far as this file goes I _could_ split it into two: a file that explains details on how things are implemented (`details.md`, `implementation.md`, something else ?) and one about obfuscation (`obfuscation.md`) but that then that also changes the way the judges' remarks looks and I quite like how it looks ...

**Comment by SirWumpus** [comment]
Who is the source of the `deobfuscated` file (weak name)?  Was it supplied by the entry author or 3rd party?  

  - By a 3rd party I suppose you can take liberties to rename stuff, but splitting information into more files seems wrong.  You're moving into the area of being an editor and possibly changing the writing teaching focus.  Also creating more work.  What if different sections of a document make reference to "see above" or "see below" but now those are in different files.
 
  - By the author I would NOT change how they presented and name their supplemental files; possible clashes and confusion with their personal repos.  In almost all my spoiler directories I have support files I used to transform a program into `prog.c`.  Some I have created C89 versions from the original K&R C and supply both.

  - If `prog.c` is the obfuscated entry, then maybe name the deobfuscated version `program.c` (original version needs more "ram" ;-) )
 
Again I repeat an earlier missive:

> I'd suggest NOT building (or cleaning) spoiler/ subdirectories (ohh, just remembered contrib/ is a fav of many, might be better than spoiler/) as part of the build system. These are supplemental to the winner and provided for learning. Leave them to the viewers to play with IMO. I wouldn't fuss with the extra stuff- just provide as-is.

I don't see an issue with `spoiler/` as a directory name, its commonly used and well understood.  If you're still searching for a name `Answers/`, `Bonus/`, `Contrib/`, `Exrta/`, `Other/`, `Supplemental/`.

You can tell we're all programmers here.  I think it takes a programmer to understand the power of what is in a name of a variable, function, file, object, etc.  The original author would have put thought into it to some degree which should be respected.

Like programmers working on an existing project, we should maintain the "coding & whitespace style" (or editorial style of documentation) of the original, no matter how much we want to scratch that itch to reformat.  (I remember being told off about such things in my first professional job after uni.)

**Comment by xexyl** [comment]
Okay looking at 2020/ferguson1 again the file name obfuscation.md could work okay .. it all is about obfuscation, some indirectly and others directly. I'll do that for this one but for the others I do not know ..

**Comment by SirWumpus** [comment]
>  I think the idea of not building or cleaning these versions has some merit. 

I'm assuming that a viewer is cloning the winners repo (or portion thereof) from GitHub.  If the viewer really wants a pristine tree, `git clean -f` will do the trick.  So there are other ways to clean the tree without having to fuss with nested makefiles.

**Comment by xexyl** [comment]
> > I think the idea of not building or cleaning these versions has some merit.
> 
> I'm assuming that a viewer is cloning the winners repo (or portion thereof) from GitHub. If the viewer really wants a pristine tree, `git clean -f` will do the trick. So there are other ways to clean the tree without having to fuss with nested makefiles.

Well that's another matter entirely and it's not necessarily the case as there are tarballs to download too. The `make clean` and `make clobber` rules are important but whether the `spoiler` code should be compiled or not I guess depends on if it offers any functional difference from the entry. Some is already set in alt code though so it might not be worth to worry about changing them all back .. but whether any additional ones added in the future should be compiled is another matter.

**Comment by xexyl** [comment]
2020/ferguson1 almost done. Would have been done earlier if I had updated the entry_text properly first. I could use sgit(1) of course but just running the script I wrote ... will be done soon. The rest I can look at later. I hope to finish other things here today and I can worry about this issue later after discussion.

**Comment by xexyl** [comment]
Commit https://github.com/ioccc-src/temp-test-ioccc/pull/2526/commits/e5bcf2ee25a17b43bf72337f4c07bd60ec7fc9e2 fixes 2020/ferguson1 .. but many others probably need to be corrected, I am afraid. Discussion required first.

Will be going afk very shortly but when back I hope to look at the other stuff: where I left off yesterday.

**Comment by SirWumpus** [comment]
This is going off-topic slightly (maybe spin a separate issue/discussion for it).

> ... as there are tarballs to download too. 

Why?  Seems like lots of extra work on top of what `git` already provides.  Viewers can always use `git sparse-checkout` to grab specific directories, rather than the whole repo.  Seems like maintaining tarballs along side what git does it a lot of fuss.

What is sgit(1)?

**Comment by xexyl** [comment]
> This is going off-topic slightly (maybe spin a separate issue/discussion for it).
> 

Somewhat related :-)

> > ... as there are tarballs to download too.
> 
> Why? Seems like lots of extra work on top of what `git` already provides. Viewers can always use `git sparse-checkout` to grab specific directories, rather than the whole repo. Seems like maintaining tarballs along side what git does it a lot of fuss.
> 

Convenience. See [FAQ 3.20 - How do I download individual winning entries or all winning entries of a given year?](https://ioccc-src.github.io/temp-test-ioccc/faq.html#faq3_20).

> What is sgit(1)?

A very useful tool I wrote that lets one run `sed` on all (or select) files in a git repo .. with a lot of really nice features. See [sgit repo](https://github.com/xexyl/sgit). Landon has actually used it in another repo (his calc repo if I'm not very much mistaken). 

Only relevant here in that it made it easier to update things. Of course one must be careful to edit the right things only but that always goes.

**Comment by lcn2** [comment]
> This is going off-topic slightly (maybe spin a separate issue/discussion for it).
> 
> ... as there are tarballs to download too.
> Why? Seems like lots of extra work on top of what git already provides. Viewers can always use git sparse-checkout to grab specific directories, rather than the whole repo. Seems like maintaining tarballs along side what git does it a lot of fuss.

We considered dropping the compressed tarballs altogether in favor of people just using `git(1)` (and things like `git sparse-checkout`).  However we received pushback and so we relented to providing compressed tarballs for each entry and for a given year to make it easy for non-git users to download entries for a new IOCCC year.

We did draw the line at not providing a tarball for everything.  For everything, for multiple years, for almost everything they will need to use git / clone the repo.

**Comment by xexyl** [comment]
> > This is going off-topic slightly (maybe spin a separate issue/discussion for it).
> > ... as there are tarballs to download too.
> > Why? Seems like lots of extra work on top of what git already provides. Viewers can always use git sparse-checkout to grab specific directories, rather than the whole repo. Seems like maintaining tarballs along side what git does it a lot of fuss.
> 
> We considered dropping the compressed tarballs altogether in favor of people just using `git(1)` (and things like `git sparse-checkout`). However we received pushback and so we relented to providing compressed tarballs for each entry and for a given year to make it easy for non-git users to download entries for a new IOCCC year.
> 
> We did draw the line at not providing a tarball for everything. For everything, for multiple years, for almost everything they will need to use git / clone the repo.

FWIW I'm glad the tarballs are there. That might seem odd as I have multiple copies of it now but it's a nice to have feature for a smaller download. Not necessary with 8TB but even so the repo is quite large and some people might not want to download it all - or might not want to get into git or its workings.

I mean after all, who likes to deal with gits? :-)

**Comment by xexyl** [comment]
Now as far as the name 'deobfuscation' for file names I think it is a mistake to do it all this way without considering the name and I think it's a mistake to make it the default name.

In the case of 2020/ferguson1 (thanks for the fix to index.html that I missed btw) obfuscation works as a name because it was all about the obfuscation (indirectly but also directly). But let's consider a different one of 2020: endoh2. That directory has some commented code but not much and what it really is, the purpose of that directory, is to show how he formatted the code. The judges' remarks also ask the question:

```
With the source from the [deobfuscation](deobfuscation/index.html) directory could you make the code
render for a larger terminal window?  How about adding more glyphs?

```

but it used to read, of course, `spoiler`, instead of the change. But this question seems odd: if it's deobfuscated (which it's not really) then shouldn't it be easy to make the changes? 

Now in this case the question of whether 'spoiler' is the right word is a valid question. But I don't think 'deobfuscation' is the right word either. Yusuke had it as 'spoiler' and I think it might be, as @SirWumpus said, right to keep it that way.

Okay there were some cases where I renamed files, sometimes maybe not necessary (as it turns out it wouldn't be but these were done before the manifest and how to view - GitHub or download - was figured out). 

One might not have been necessary at all: a matter of semantics only and I have on more than one occasion thought of reverting that change for that reason ('olympic' to 'olympics').  The question is whether it refers to the games or is the games. The problem is determining this though as I don't have a shortwave radio .. or if I do (no idea now) it certainly cannot work with this. So with that in mind I've renamed it to what it was but I'll have to update the manifest another time (let me rephrase that: commit the manifest changes).

But in general I think if an author provides a spoiler version it might be good to keep the name as they provide it.

A question of consistency might arise here but the thing is: different authors do things differently (of course and as we would want) so we can't really aim for consistency. In the html files that were spoiler.html I think that's a good name for what they do: typing out how it was programmed (at least that's what I gathered from it .. as you can probably imagine I never watched more than a few seconds of it). 

So if consistency of names of files cannot be or should not be the goal what should they be called? If we want to rename all versions that are less obfuscated (or something else - as referenced above not all of these files are about deobfuscation as such) where do we draw the line?

There are many entries, mine and others, that could have files renamed but which do they need to be renamed?

Also, one more thing to consider: as long as the entry_text of the manifest is correct does it really matter what the name is? And although this might not be the case (I have not verified this btw) we have to remember that some names are hardcoded into obfuscated code and changing the names could break the code. If it's just the file name we can change it to `__FILE__` (as you did for some and I did for others and maybe others did too) but this is not always the case. I distinctly remember one case where I could not use `__FILE__` actually.

Of course if these should be considered (and I think they should be, especially the paragraph right above this) then the question is how to go about changing the names back. In some cases it might be okay but I remember that a lot of the files (probably most if not all) that are now called 'deobfuscation' are actually just less obfuscated but they are not not obfuscated. And to be pedantic (sorry :-) ) deobfuscation sounds like how it might be done rather than it being done. But as noted this is not usually the case.

Just some things to think about. As this is not a high priority issue I won't worry about it for now (other than the one change above I noted) but rather let this be thought about and we can have any discussion that might be necessary or useful.

**Comment by lcn2** [comment]
Fair points raised in [comment-2204483053](https://github.com/ioccc-src/temp-test-ioccc/issues/2513#issuecomment-2204483053).  

One may wish to reconsider parts of commit 4d6009b58fa56e966818fc4b42d8741ec4f64e73, and commit 3aec80f057cf4863a41fba40e2039293694aead1, and commit 85c991620f605d3905f386d52fe3ed3a2a32fa85, and commit 5dd3311e9ef36b4570cd972b6d1ec327ae7a4c78. perhaps?

**Comment by xexyl** [comment]
Okay copied the older copy (the only good thing from the stupid mistake I made at the stupid hour I was doing it at on the fedora box .. but getting dangerously close to EOL :(( ) to the /home on the server and just rsynced to the laptop as ~/code/old-temp-test-ioccc/. The downside is a lot of other things have also changed so I'm not sure how useful it's going to be except for filenames: which admittedly are not many. Problem is there have been many other filenames changed as well since that was last pulled ...

I suppose you don't have any idea what changed, do you Landon?

**Comment by xexyl** [comment]
I figured out the spoiler versus otherwise: if nothing but the filenames and text of the words 'spoiler' and 'spoilers' were changed. I just did a pull and then checked out the commit immediately before you started.

Then i did..

```sh
git ls '*spoiler*'
git grep -il spoiler '*md'
```

to get a list of files that have the word spoiler in it (though I do recall now I think on it an upper case file) and then the list of files that have the word, case insensitive, 'spoiler' (or of course 'spoilers').

Now I can do a diff on the two directories too and see what was changed: I hope.

The question is how to go about the naming. It might be better to just change it back to what the authors had but I'm not sure yet: it seems like you might want otherwise but I think keeping the author's filenames the same (or close to the same: a few changed like spoiler1 to spoiler and some others that were not named to a way that might suggest what they are). The text is easy to fix with `sgit(1)` of course though I have to check all the cases too.

**Comment by xexyl** [comment]
Now as far as the URLs in the top comment here I think that should be addressed only after the other files are checked. The question is will the subdirectory work? And another question is should they be compiled? I think in this case at least they should not.

Another problem I just thought of: if any prog.alt.c files were changed then those are missed in the above commands.

**Comment by xexyl** [comment]
In some cases 'obfuscation' works well and probably better than 'spoiler' but certainly better than 'deobfuscation' as it describes how it is obfuscated rather than how to deobfuscate it - so I have changed those though I have quite a few files to look at I think. That's not considering the files being renamed: that will come after the text .. I think.

**Comment by xexyl** [comment]
Making good progress ... my `lessdiff` script is proving very useful. I will be going afk for a bit soon though .. and not long after that I will have to leave for a while. I hope to at least finish this task today, however.

**Comment by xexyl** [comment]
Well I just went through 2004/hibachi and I think that's a good place to stop for now .. will do more later. As a reminder to myself the command I used:

```sh
lessdiff -r . ../xexyl-temp-test-ioccc
```

Back later.

**Comment by xexyl** [comment]
Okay I must leave now (or very shortly) but when I'm back I hope to fix the rest of the changes. That depends on how I feel but if I do it will require some effort and the manifest will have to be updated. I expect having to do this a few times to make sure everything is in order. I could possibly get away with the quick_readme2index rule but to be safe I'll do the www one - which of course a fair amount more time to complete.

I guess after that I will not be up to anything else but we'll see. The weather is not helping matters of course.  Hopefully no power outage today.

Anyway back later and hopefully this first part will be done. Then I will look at the information from @SirWumpus and after that go back to the other tasks of #2006.

**Comment by lcn2** [comment]
We believe we have addressed all of the current questions that still need answering at this time. If we've missed something or something else needs to be clarified, please ask again.

**Comment by xexyl** [comment]
Almost done with the fixes .. have to update manifest but first going to run test on the links. 

What would be very helpful is if you could update the `DATA` variables in the Makefile if necessary: I'm not sure it affects all entries.

Seeing a list of renaming when merging would make it easier I think. BUT I wonder what the purpose is anyway? Do we really need the compilation of the programs to depend on `DATA`? Even so the variable should be updated to be consistent with filenames.

**Comment by xexyl** [comment]
.. that being saidI must go afk .. quite a while I think: at least an hour I think.

**Comment by xexyl** [comment]
No missing links: at least not direct to repo link: if any are missing in `%%REPO_URL%%` or similar I don't know. For reference for the `DATA` variable this might be useful for filename changes and maybe I can take care of it looking at this list:

	renamed:    ../../1995/leo/deobfuscation.html -> ../../1995/leo/secret.html
	renamed:    ../../1995/leo/deobfuscation.md -> ../../1995/leo/secret.md
	renamed:    ../../2004/hibachi/src/localhost/CHANGES-DEOBFUSCATED.TXT -> ../../2004/hibachi/src/localhost/CHANGES-SPOILER.TXT
	renamed:    ../../2012/omoikane/deobfuscation.html -> ../../2012/omoikane/obfuscation.html
	renamed:    ../../2013/misaka/deobfuscation.html -> ../../2013/misaka/obfuscation.html
	renamed:    ../../2014/endoh1/deobfuscation.html -> ../../2014/endoh1/quine-qr.html
	renamed:    ../../2014/endoh1/deobfuscation.md -> ../../2014/endoh1/quine-qr.md
	renamed:    ../../2014/sinon/deobfuscation.html -> ../../2014/sinon/obfuscation.html
	renamed:    ../../2015/burton/deobfuscation.html -> ../../2015/burton/obfuscation.html
	renamed:    ../../2015/burton/deobfuscation.md -> ../../2015/burton/obfuscation.md
	renamed:    ../../2015/howe/deobfuscation/Makefile -> ../../2015/howe/deobfuscated/Makefile
	renamed:    ../../2015/howe/deobfuscation/err.h -> ../../2015/howe/deobfuscated/err.h
	renamed:    ../../2015/howe/deobfuscation/prog.c -> ../../2015/howe/deobfuscated/prog.c
	renamed:    ../../2015/yang/deobfuscation.html -> ../../2015/yang/obfuscation.html
	renamed:    ../../2018/algmyr/deobfuscation.png -> ../../2018/algmyr/spectrogram.png
	renamed:    ../../2018/yang/deobfuscation.html -> ../../2018/yang/obfuscation.html
	renamed:    ../../2019/adamovsky/deobfuscation.c -> ../../2019/adamovsky/deobfuscated.c
	renamed:    ../../2019/yang/deobfuscation.html -> ../../2019/yang/obfuscation.html
	renamed:    ../../2020/endoh2/deobfuscation/code.c -> ../../2020/endoh2/obfuscation/code.c
	renamed:    ../../2020/endoh2/deobfuscation/gen.rb -> ../../2020/endoh2/obfuscation/gen.rb
	renamed:    ../../2020/endoh2/deobfuscation/glyphs.txt -> ../../2020/endoh2/obfuscation/glyphs.txt
	renamed:    ../../2020/endoh2/deobfuscation/index.html -> ../../2020/endoh2/obfuscation/index.html
	renamed:    ../../2020/endoh2/deobfuscation/index.md -> ../../2020/endoh2/obfuscation/index.md
	renamed:    ../../2020/endoh2/deobfuscation/prog.c -> ../../2020/endoh2/obfuscation/prog.c
	renamed:    ../../2020/endoh2/deobfuscation/tmpl.txt -> ../../2020/endoh2/obfuscation/tmpl.txt
	renamed:    ../../2020/yang/deobfuscation.html -> ../../2020/yang/obfuscation.html

..but that can come later .... will start the manifest update script and then will be afk a while .. I do have some other things to take care of when I'm back, too, I'm afraid, but I might be able to do it in between things.

**Comment by xexyl** [comment]
Actually I'll do this when I'm back (here just a moment) and then do a commit: everything else in order I think. I do have an idea for something else that belongs in another issue but I can't say it now as I have to leave again ... hopefully remember it.

**Comment by xexyl** [comment]
With commit https://github.com/ioccc-src/temp-test-ioccc/pull/2541/commits/1ac055fa85daccab8e09a86f6a7681eef69afea2 the wording situation should be in a much better state.

The next step with this issue is to add the files linked to in the top comment. I am not sure if I will get to that today or not: it's one of those days. But I would like to so that I can then get to seeing how this looks. There is something I need to do though and there is something I should do and I'll be doing those first. I am sure I will manage to do more here today but this is a good change.

And I took care of the `DATA` variable. Some were not even used (empty) so I didn't worry about those. Some I added files (since they already had some). I still wonder if:

```makefile
# alternative executable
#
alt: data ${ALT_TARGET}
        @${TRUE}

${PROG}.alt: ${PROG}.alt.c
        ${CC} ${CFLAGS} $< -o $@ ${LDFLAGS}
```

is a good idea though. In that case it was once `prog` also but I removed it: it's in 2020/ferguson1. But there are other entries that are not mine that also have this and it means that if a file is missing or is renamed and the Makefile is not updated it fails to compile. In fact when this repo was created (or the winner one) and the files were renamed from .text to .markdown (or whatever they were .. not sure if all were immediately named .md but I do know that hint files were renamed by me to README.md files at your request) it prevented 2020/ferguson1 from compiling and for a while I had on my site a tarball with the fixes.

**Comment by lcn2** [comment]
> With commit 1ac055f the wording situation should be in a much better state.

We agree that there are a number of improvements to in that commit. 

There are a few places what the term "_spoiler_" was re-introduced that we need to ponder. 

Our initial guess is that we might accept the pull request as is, and reconsider a few of the changes in a later commit.

We need to study the effect of this pull request some more.

**Comment by xexyl** [comment]
> > With commit [1ac055f](https://github.com/ioccc-src/temp-test-ioccc/commit/1ac055fa85daccab8e09a86f6a7681eef69afea2) the wording situation should be in a much better state.
> 
> We agree that there are a number of improvements to in that commit.
> 
> There are a few places what the term "_spoiler_" was re-introduced that we need to ponder.
> 
> Our initial guess is that we might accept the pull request as is, and reconsider a few of the changes in a later commit.
> 
> We need to study the effect of this pull request some more.

I just gave you there two examples where it cannot possibly work otherwise. There are others. The problem with it is that many of them are not about obfuscation or deobfuscation. In many cases the deobfuscation does not work well at all. And the ...we'll stop spouting spoilers... one would not work with deobfuscation information either. it has nothing to do with deobfuscating the entry. It simply is a summary of what it is. So there are more examples where it would not work.

I think to be that against the word spoiler is the wrong way to go about it. It works better than the proposed suggestions in some cases and in some cases it's the right word. In other cases there are no better words that I can think of off the top of my head but seeing the changes in the web pages might help figure it out.

**Comment by xexyl** [comment]
Anyway I must go for now .. will likely be a couple hours or so. Not sure what I can do after that but it also kind of depends on this.

**Comment by xexyl** [comment]
Quickly .. just thought of an alternative for stop spouting spoilers. Summaries. Brief descriptions. But it doesn't have quite the same feeling. But it could be done. Afk now.

**Comment by lcn2** [comment]
> Quickly .. just thought of an alternative for stop spouting spoilers. Summaries. Brief descriptions. But it doesn't have quite the same feeling. But it could be done. Afk now.

That is a very interesting idea 💡 @xexyl.  

As we mentioned elsewhere, our inclination is to accept the pull request as it is now (after studying it first) and then ponder suggestions such as the one you made above.

**Comment by xexyl** [comment]
> > Quickly .. just thought of an alternative for stop spouting spoilers. Summaries. Brief descriptions. But it doesn't have quite the same feeling. But it could be done. Afk now.
> 
> That is a very interesting idea 💡 @xexyl.
> 
> As we mentioned elsewhere, our inclination is to accept the pull request as it is now (after studying it first) and then ponder suggestions such as the one you made above.

Okay .. I hope you do accept it but if not please let me know what you want changed so I can fix it as it would be a lot of work to do it all again and not nearly as much work to fix.

Thanks.

**Comment by xexyl** [comment]
Just took a few minutes to look at the first one .. one file had a few compilation errors. Resolved those. Updated Makefile (to be more GNU like so it would compile to the right file name). Updated README.md. Will check the index.html file. Will update manifest too. But I won't do the manifest update until the other entries are looked at as well.

**Comment by xexyl** [comment]
Okay before I do the manifest update ... a general question. Should these become `alt` versions in the entry directory or should they be put in a subdirectory `deobfuscated`? Both ways are used in the repo in different entries. What if there already is an alt version?

If there's any use of a try script (like there is for 1992) then it would be better to have it as an alt file, assuming we want it to compile, but if not it might be better in a subdirectory?

The discussion of whether or not to compile it if it has no functional difference comes up. On the one hand it seems like it might not be useful but on the other it might help people explore how the original one worked if they can compile both (that's how I'm looking at it now).

So I think what I'll do is I'll get these to compile and then before I do a commit (or update the manifest) we can discuss this. That will delay completing this issue but hopefully not by very long.

**Comment by xexyl** [comment]
But for now I have to go afk a while ... and it very possibly will be a slow day. Feeling off. I hope that will change but it's possible. Still I guess I will be able to get these all in today even if it has to be changed somewhat.

**Comment by xexyl** [comment]
> Okay before I do the manifest update ... a general question. Should these become `alt` versions in the entry directory or should they be put in a subdirectory `deobfuscated`? Both ways are used in the repo in different entries. What if there already is an alt version?
> 
> If there's any use of a try script (like there is for 1992) then it would be better to have it as an alt file, assuming we want it to compile, but if not it might be better in a subdirectory?
> 
> The discussion of whether or not to compile it if it has no functional difference comes up. On the one hand it seems like it might not be useful but on the other it might help people explore how the original one worked if they can compile both (that's how I'm looking at it now).
> 
> So I think what I'll do is I'll get these to compile and then before I do a commit (or update the manifest) we can discuss this. That will delay completing this issue but hopefully not by very long.

Okay obviously in some they have to be a subdirectory as there are other files like a test Makefile and so on. So maybe because of that to make the entries consistent they should all be a subdirectory. I'll do that .. when back.

**Comment by xexyl** [comment]
I have come up with what I think is a nice way to present the files. The `am` one was giving me some issues but I got it sorted. It does change the way the Makefile `alt` rule looks but it keeps the `deobfuscated/` directory clean. I will be leaving for a bit soon possibly but when back I will work on the next one. After the files are in I can update the README.md files and then I can update the manifest.

**Comment by SirWumpus** [comment]
What challenge did `am` give you?  Asking for a friend.

**Comment by xexyl** [comment]
> What challenge did `am` give you? 

It was not `am` itself but integrating it into the Makefile in a way that works with the `alt` model.

> Asking for a friend.

Of course you are ... yourself? Or maybe your √-1 friend? :-)

**Comment by xexyl** [comment]
.. although I wonder if this is desired, to separate it like this. I don't really know. I think it helps with identifying it as being deobfuscated code but maybe it should just be `alt` in the entry directory. The main reason I have it in the subdirectory is you have provided Makefiles and the main Makefile uses them for the alt build.


Well we'll see I guess. It can be changed in how it's organised later.

**Comment by xexyl** [comment]
1991, 1992 and 1993 done except for the manifest. Will work on the next ones now.

**Comment by xexyl** [comment]
Hmm ... well hibachi ought to be fun!

I am not sure how to best go about this. I think for starters I'll just add the code to the subdirectory and update the manifest and README.md file to note it. Then it can be decided how to best approach building it.

**Comment by xexyl** [comment]
> * 2005 https://github.com/SirWumpus/ioccc-mynx (original http: version)


Question for you @SirWumpus. I see 'original http: version'. Does that mean you have a `https` version? If you do would you share it? I had added alt code that scans (but only scans) in case someone wanted to try and get a command line that might work out (what that would be I don't know) but obviously if you have a version that works with https that would be better and I would happily add it for you. As one of the files says: without this it might just have to be an entry that can be enjoyed for what it is, without it working for many websites.

Let me know and we can go from there.

**Comment by SirWumpus** [comment]
I never had an `https` version.  I think someone did a conversion.  Otherwise I can look into it this weekend.

**Comment by xexyl** [comment]
... actually on second thought I'm not going to do `2004/hibachi` right away. There are too many questions that have to be answered first: should it use the `configure` script is a good example. But then it also has the Makefile template and it would all have to be updated to not have the `-spoiler` part.

**Comment by xexyl** [comment]
> I never had an `https` version of `mynx`. I think someone did a conversion. Otherwise I can look into it this weekend.

That would be excellent if you would like to do that! It would mean the entry could be used with the IOCCC website again, for example. Of course it's been a long time so maybe it's harder to get the same ... and setting up https isn't exactly fun (or so I think - I haven't enjoyed it with sockets anyway, getting encryption to work, usually opting for stunnel).

**Comment by SirWumpus** [comment]
One of the things about `hibachi` as winner, was that it supplied a `configure` script, so it really should appear in the build process.  I think you want an `ioccc.mk` and the `hibachi` `configure` and `Makefile.in`.  Have the root level Makefile either invoke `hibachi/ioccc.mk` or have the root go directly to the configure/makefile.  <baby dinosaur>Gotta love me!</>.

**Comment by xexyl** [comment]
Okay @lcn2 I have 1991, 1992 and 1993 sorted though it might be that we want to reorganise how these deobfuscated versions are built.

However the question is how to go about hibachi and maybe mynx with their configure scripts and such. One thought is that I could run `./configure` and add the generated files and then have the entry's Makefile updated along the lines of how the original entry build works.

In one case at least one would have to remove `-spoiler` and just have the name `hibachi` (not sure if `mynx` has that form yet) but this should not be a problem.

**Comment by xexyl** [comment]
> One of the things about `hibachi` as winner, was that it supplied a `configure` script, so it really should appear in the build process. I think you want an `ioccc.mk` and the `hibachi` `configure` and `Makefile.in`. Have the root level Makefile either invoke `hibachi/ioccc.mk` or have the root go directly to the configure/makefile. <baby dinosaur>Gotta love me!</>.

Yeah you're right. Hence the comment I just posted. It might be the way to go about it but it has to be discussed first I think. Same possibly with `mynx`.

**Comment by xexyl** [comment]
> Okay @lcn2 I have 1991, 1992 and 1993 sorted though it might be that we want to reorganise how these deobfuscated versions are built.
> 
> However the question is how to go about hibachi and maybe mynx with their configure scripts and such. One thought is that I could run `./configure` and add the generated files and then have the entry's Makefile updated along the lines of how the original entry build works.
> 
> In one case at least one would have to remove `-spoiler` and just have the name `hibachi` (not sure if `mynx` has that form yet) but this should not be a problem.

These same questions apply to `mynx` too as I thought. Or maybe that's `as I feared` :-)

**Comment by SirWumpus** [comment]
You got to love winners that make life ... interesting.

Hibachi Sidebar fact:  The code took about two days to write, while animated gif logo took about 5 days to design each frame and build up the .gif.

**Comment by SirWumpus** [comment]
BTW a `mynx` with `https` support would be need a `configure` script :) to get all the libs in correct.

**Comment by xexyl** [comment]
> You got to love winners that make life ... interesting.

Yes indeed. I think that all winning entries have this in some form or another but the configure scripts were a unique thing for those two to be sure.

But I have to say that whatever the answers to these questions are, they will probably not be as complicated as fixing entries to not use `-traditional-cpp` - for one example kind of fix I had to do (unfortunately clang does not support it). Still those two entries will complicate matters.

The nice thing is that I already did 2015/howe so I can now just update the manifest and once it has been decided how to do the other two those can be done and then this issue can hopefully be closed.



> 
> Hibachi Sidebar fact: The code took about two days to write, while animated gif logo took about 5 days to design each frame and build up the .gif.

Interesting. I did have to fix links - some probably to refer to the Internet Wayback Machine and maybe other issues besides in the html files. 

I have had a similar experience but a different kind. I guess we all have: the things that you might think would take longest actually go relatively quickly but the simplest (sounding) things take a lot longer.

**Comment by xexyl** [comment]
> BTW a `mynx` with `https` support would be need a `configure` script :) to get all the libs in correct.

Well considering you already supplied a configure script ... :-)

**Comment by xexyl** [comment]
The `c89` file is to work for C89 or is it more like to look like c89? At a quick glance it seems to use ANSI style functions so I'm not sure.

**Comment by SirWumpus** [comment]
> The c89 file is to work ...

Which entry you referring too?  In my spoilers `ae-c89.c` and `ag-c89.c` (I doubt I got around to`am-c89.c`) are conversions from the original K&R source to something that will compile with `-std=c89` or higher.  New compilers complain about K&R C unless you turn on an `-std` for it (gcc supports K&R just, but I don't think `clang` does).

**Comment by xexyl** [comment]
> > The c89 file is to work ...
> 
> Which entry you referring too? In my spoilers `ae-c89.c` and `ag-c89.c` (I doubt I got around to`am-c89.c`) are conversions from the original K&R source to something that will compile with `-std=c89` or higher. New compilers complain about K&R C unless you turn on an `-std` for it (gcc supports K&R just, but I don't think `clang` does).

Right but I got the other version to compile too - though with a few fixes necessary. But that answers the question: it's a conversion from the original K&R. But this brings up the question of what the original .. aha. I see what you mean, maybe: that might explain why I had to fix the version that was not `c89` but the `c89` worked. Is that so?

**Comment by xexyl** [comment]
> > The c89 file is to work ...
> 
> Which entry you referring too? In my spoilers `ae-c89.c` and `ag-c89.c` (I doubt I got around to`am-c89.c`) are conversions from the original K&R source to something that will compile with `-std=c89` or higher. New compilers complain about K&R C unless you turn on an `-std` for it (gcc supports K&R just, but I don't think `clang` does).

As far as `clang` and `-std=` it does support it but it certainly is a much more strict compiler and it is the reason that many entries had to be fixed, especially as macOS would not work even if you invoke it as `gcc` (it is clang): the default install does not have gcc; you have to install it with MacPorts or Homebrew or compile it yourself.

**Comment by SirWumpus** [comment]
> ... not c89 but the c89 worked. Is that so?

Yep. The C89 versions should compile without errors.

> As far as clang and -std= it does support it but 

`clang` has a `-std` option but I don't not think tit has a K&R standard setting, eg. `-std-knr`.  I tried looking last year.

`gcc` will detect K&R and I recall there is some option to enable to silence some errors or warnings concerning the K&R function declarations.

**Comment by xexyl** [comment]
> > ... not c89 but the c89 worked. Is that so?
> 
> Yep. The C89 versions should compile without errors.
> 

They did fine yes. I figured it was the opposite purpose which is why it surprised me.

> > As far as clang and -std= it does support it but
> 
> `clang` has a `-std` option but I don't not think tit has a K&R standard setting, eg. `-std-knr`. I tried looking last year.

It might not have a `knr` to it, no. If you didn't see one then it's extremely unlikely it was added. It has some defects or at leat some features that break older code that shouldn't really break. Some code obviously does break over time but some I think should not break. The C committee of course is not a help there.

> 
> `gcc` will detect K&R and I recall there is some option to enable to silence some errors or warnings concerning the K&R function declarations.

Yes `gcc` is far more forgiving in things that `clang` is against - including things that seem questionable. Fortunately it was easy to fix the other one for `clang`. Just a couple minor things.

**Comment by xexyl** [comment]
... updating manifest .. will take a bit of time but when done I will do a pull request. Then it's a matter of figuring out hibachi and mynx. After that this issue can be closed (I guess). Though I will probably not check the links until tomorrow as I have been at it too long.

**Comment by SirWumpus** [comment]
I've added an `am-c89.c` to my personal repo spoiler directory.   You probably don't need it since it was actually pretty small and straight forward K&R to C89 clean-up.

https://github.com/SirWumpus/ioccc-am/blob/master/spoiler/am-c89.c

**Comment by xexyl** [comment]
> I've added an `am-c89.c` to my personal repo spoiler directory. You probably don't need it since it was actually pretty small and straight forward K&R to C89 clean-up.
> 
> https://github.com/SirWumpus/ioccc-am/blob/master/spoiler/am-c89.c

Probably not needed indeed since I got the other one to compile. But maybe to be consistent it should be added. What do you think, Landon? Should I add it? If yes I can do that .. tomorrow. Should be done updating the manifest soon - I hope. The spreadsheet and CSV file have been updated some time ago and just waiting on the scripts to finish.

**Comment by xexyl** [comment]
You might enjoy something too Anthony that I put in the `entry_text` in your `sleep.c` file:

```
author provided code for those who need a nap
```

so the link in the index.html will look like:

![Screenshot 2024-07-05 at 12 40 44](https://github.com/ioccc-src/temp-test-ioccc/assets/53008573/732e81ac-d84b-4c8e-b5b5-3145b1da4c36)

I did this because I did not know the real purpose of it, it amused me and I thought you and Landon would find it amusing too. But I won't note it in the changelog .. let people discover it for themselves.

**Comment by SirWumpus** [comment]
The `ioccc-am/spoiler/sleep.c` is used when building using WatCom C back in the day.  I use to build my entries for DOS, AtariST, SysV, and Solaris to test portability.   WatCom C on DOS didn't provide any unix/posix support functions then.

**Comment by SirWumpus** [comment]
In theory you could probably drop the `sleep.c`, since I doubt anyone will try to build with DOS 3.3 these days or have WatCom C.

**Comment by xexyl** [comment]
> The `ioccc-am/spoiler/sleep.c` is used when building using WatCom C back in the day. I use to build my entries for DOS, AtariST, SysV, and Solaris to test portability. WatCom C on DOS didn't provide any unix/posix support functions then.

Ah. Well it's a fun Easter egg for people to find now. AtariST was the first OS I used. I do seem to recall seeing something about you testing in some of those systems before but unless it was in a README file I don't know where. I have fond memories of SunOS/Solaris too but it's been years since I last used it. I have a mate who had an UltraSPARC and we used to play with it (though for me from a distance ... that was back in the days before ssh).

**Comment by xexyl** [comment]
> In theory you could probably drop the `sleep.c`, since I doubt anyone will try to build with DOS 3.3 these days or have WatCom C.

I doubt it too but it's a fun addition for people to find and it's already in so why not?

**Comment by xexyl** [comment]
Commit https://github.com/ioccc-src/temp-test-ioccc/pull/2543/commits/a4fb0a8c8ca85751e2069a7e814a282894a66756 takes care of 199[1-3]/ant. The other two can be decided later (it seems I did 2015 yesterday). What can also be discussed is if the way the build of alt code should be reorganised: it might need it but I'm not sure.

I will look at the index.html files tomorrow .. been at this too long today and I'll be leaving soon most likely.

**Comment by xexyl** [comment]
I am not convinced the look of the `Alternate code` and the reference of `deobfuscated/` looks good in these three entries and probably the same in 2015 (or at least possibly).

However I do not feel up to fixing it today. Tomorrow should be better. But we still need to discuss the other two years, the ones with `configure`, before I can do those. Also if you are not sure of using the `Makefile` in the subdirectory we should talk about that (I was unsure but it does keep it clean). I don't like the directory name `deobfuscated` that much but I can't think of another word if `spoiler` is not allowed. I mean it makes sense in these entries but not sure how it looks. Still the way the presentation of it in index.html looks is I think bad.

But I will have to worry about this tomorrow - hopefully with discussion already. I will reply to some comments in another issue and then I will be doing other things for the rest of the day.

**Comment by lcn2** [comment]
With commit 2e2ae5f132f2e4f7f60ad8c275038c8db506c7ec we add notices to those who wish for a greater challenge.

We believe this achieves a better convergence to our intent with the fixes and improvements that were made by @xexyl in response to our previous but not very good attempt.

# The details:

We add the following level-3 header to help people identify text
they may wish to skip if they wish for a greater challenge:

```
### NOTICE to those who wish for a greater challenge
```

After that notice we add:

```
**If you want a greater challenge, don't read any further**:
just try to understand the program via the source.

If you get stuck, come back and read below for additional hints and information.
```

And sometimes follow with:

```
### How this entry works:
```

Or sometimes follow with:

```
### What this entry does:
```

Where the alt code impacts a "greater challenge", we add instead:

```
**If you want a greater challenge, don't read any further**:
just try to understand the program without looking at the Alternate code.

If you get stuck, come back and look at the Alternate code.
```

And where a given command impacts a "greater challenge", we add instead:

```
**If you want a greater challenge, don't read any further**:
just try to understand the program without running the command below.

If you get stuck, come back and run the following command:
```

**Comment by xexyl** [comment]
> With commit 2e2ae5f132f2e4f7f60ad8c275038c8db506c7ec we add notices to those who wish for a greater challenge.
> 
> 
> 
> We believe this achieves a better convergence to our intent with the fixes and improvements that were made by @xexyl in response to our previous but not very good attempt.
> 
> 
> 
> # The details:
> 
> 
> 
> We add the following level-3 header to help people identify text
> 
> they may wish to skip if they wish for a greater challenge:
> 
> 
> 
> ```
> 
> ### NOTICE to those who wish for a greater challenge
> 
> ```
> 
> 
> 
> After that notice we add:
> 
> 
> 
> ```
> 
> **If you want a greater challenge, don't read any further**:
> 
> just try to understand the program via the source.
> 
> 
> 
> If you get stuck, come back and read below for additional hints and information.
> 
> ```
> 
> 
> 
> And sometimes follow with:
> 
> 
> 
> ```
> 
> ### How this entry works:
> 
> ```
> 
> 
> 
> Or sometimes follow with:
> 
> 
> 
> ```
> 
> ### What this entry does:
> 
> ```
> 
> 
> 
> Where the alt code impacts a "greater challenge", we add instead:
> 
> 
> 
> ```
> 
> **If you want a greater challenge, don't read any further**:
> 
> just try to understand the program without looking at the Alternate code.
> 
> 
> 
> If you get stuck, come back and look at the Alternate code.
> 
> ```
> 
> 
> 
> And where a given command impacts a "greater challenge", we add instead:
> 
> 
> 
> ```
> 
> **If you want a greater challenge, don't read any further**:
> 
> just try to understand the program without running the command below.
> 
> 
> 
> If you get stuck, come back and run the following command:
> 
> ```

Well without having looked at it just yet that does sound like an improvement. 

Thanks! The only other question is the directory name. But I will be booting up the laptop next and I can take a look.

**Comment by SirWumpus** [comment]
> The only other question is the directory name.

`Dragons/` as in `Beyond here be dragons.`

**Comment by xexyl** [comment]
> > The only other question is the directory name.
> 
> `Dragons/` as in `Beyond here be dragons.`

Sorry but would that be Smaug? Scatha? Ancalagon? When you're talking to Landon and me about dragons 🐉 you have to be more specific!

**Comment by SirWumpus** [comment]
> When you're talking to Landon and me about dragons 🐉 you have to be more specific!

`Puff` https://www.youtube.com/watch?v=Y7lmAc3LKWM

**Comment by xexyl** [comment]
> > When you're talking to Landon and me about dragons 🐉 you have to be more specific!
> 
> `Puff` https://www.youtube.com/watch?v=Y7lmAc3LKWM

Hahaha .. well if you're going for magic dragons why not [Piff the Magic Dragon](https://www.youtube.com/user/piffthemagicdragon)?

Anyway I thank you for elaborating!

**Comment by xexyl** [comment]
Updating manifest now (`.entry.json` stage) for 1991, 1992, 1993 and 2015. However there still is the question of what to do about 2004 and 2005: as they have a `configure` script and Makefile templates. How do you want to deal with this, Landon? I can't add these files to finish this issue until this is decided.

**Comment by xexyl** [comment]
Question for you @SirWumpus.

In 2015/howe you wrote a summary for support files. You left out a summary for avgtime.sh though. Do you wish to provide one? The entry_text in the manifest has: `support script` .. but that could be improved too. I could come up with something, maybe, but I want to offer you the opportunity first.

**Comment by xexyl** [comment]
Commit https://github.com/ioccc-src/temp-test-ioccc/pull/2548/commits/92139c55d75dd60e7ddc2e6cf76e446d561285f3 updates the 1991-1993 and 2015 entries plus the manifest for a couple other entries (as the entry_text was made the same wording - unobfuscated).

What remains for this issue is 2004 and 2005 entries but I don't know how we want to have those organised. Please advise, Landon, when you can, so that we can hopefully get this issue resolved soon.

**Comment by SirWumpus** [comment]
> You left out a summary for avgtime.sh 

Was only intended to help test basic performance of `npdif` (mine) verses the system `diff(1)` (typically GNU).  So feel free to apply your wit "in a timely manner".

**Comment by xexyl** [comment]
> > You left out a summary for avgtime.sh
> 
> Was only intended to help test basic performance of `npdif` (mine) verses the system `diff(1)` (typically GNU). So feel free to apply your wit "in a timely manner".

Thanks .. and great pun! I put in the manifest `report average time running a command N times` and in the README.md file `Run time(1) N times on a command to determine average runtime.`

I think that works well enough. Unless you think something better can be contrived?

**Comment by xexyl** [comment]
Oh .. I just remembered the ideas you had about text to have about for a challenge etc. I did not add that to the entries I moved the alt code in. I will be finishing the four up today (except for the one with the pending question/concern) - perhaps you'd like to do this? Otherwise I can look at it tomorrow. I just feel like that issue is taking time away from this issue and it's taking longer than I was hoping :(

**Comment by SirWumpus** [comment]
> the ideas you had about text to have about for a challenge etc

Say woot?

**Comment by xexyl** [comment]
> > the ideas you had about text to have about for a challenge etc
> 
> 
> 
> Say woot?

What Landon committed. But it I wasn't clear it's because I am very tired. Can try again tomorrow.

**Comment by xexyl** [comment]
I got hibachi done .. well minus the manifest. That will be fun! 

I'll also have to do the other one too. Might as well finish that before doing the manifest.

An odd problem though with hibachi: it seems only the unobfuscated version is working. But perhaps that's due to socket options that were not used (just a guess based on my experience with sockets). Will have to try it in a bit.

**Comment by xexyl** [comment]
Working on manifest for both .. if all good then this issue can be closed (I think) - once something is decided in another issue (which is a simple fix once it's decided).

The `configure` script for hibachi was quite annoying but I got it. The mynx one has a configure for both but they're not used: I only added them because they originally were submitted. (Not my doing of not using the script .. if I had my way the other one wouldn't either though once this is resolved that will likely change: a unique thing is very much a good thing).

**Comment by xexyl** [comment]
Commit https://github.com/ioccc-src/temp-test-ioccc/pull/2551/commits/0d65af1b2bfd66ecc9120125dbb4dc1accd8d0e1 added 2004/hibachi and 2005/mynx. I have not looked at the manifest results as I haven't got to those years yet but I suspect that there won't have to be any real changes if any at all (unless it is decided that the order of files should be changed).

This means that this issue is almost resolved - just the one thing we've been discussing. But I am off for the day. Back tomorrow!

**Comment by lcn2** [comment]
Thank you @SirWumpus and @xexyl

**Comment by xexyl** [comment]
> Thank you @SirWumpus and @xexyl

You are most welcome .. though it's not completely resolved yet. Nonetheless it's about to be. I'll make a commit very soon: within minutes at most.

**Comment by xexyl** [comment]
Commit https://github.com/xexyl/temp-test-ioccc/commit/88d8a302b4eec834bee4272d5f28b374ee65c6bf should be it!

Now (or next time I look at the repo .. not sure how much more I'll do here today) I can look at the 'look and feel' issue again.

**Comment by xexyl** [comment]
Commit 88d8a302b4eec834bee4272d5f28b374ee65c6bf should be it!

Now (or next time I look at the repo .. not sure how much more I'll do here today) I can look at the 'look and feel' issue again.
